packages/frontend/component/src/ui/dnd/draggable.ts

```diff
@@ -0,0 +1,114 @@
+import { draggable } from '@atlaskit/pragmatic-drag-and-drop/element/adapter';
+import { useEffect, useMemo, useRef, useState } from 'react';
+import { flushSync } from 'react-dom';
+
+type DraggableGetFeedback = Parameters<
+  NonNullable<Parameters<typeof draggable>[0]['getInitialData']>
+>[0];
+
+type DraggableGet<T> = T | ((data: DraggableGetFeedback) => T);
+
+function draggableGet<T>(
+  get: T
+): T extends undefined
+  ? undefined
+  : T extends DraggableGet<infer I>
+    ? (args: DraggableGetFeedback) => I
+    : never {
+  if (get === undefined) {
+    return undefined as any;
+  }
+  return ((args: DraggableGetFeedback) =>
+    typeof get === 'function' ? (get as any)(args) : get) as any;
+}
+
+export interface DraggableOptions<
+  T extends Record<string, unknown> = Record<string, unknown>,
+> {
+  data?: DraggableGet<T>;
+  dataForExternal?: DraggableGet<{
+    [Key in
+      | 'text/uri-list'
+      | 'text/plain'
+      | 'text/html'
+      | 'Files'
+      // eslint-disable-next-line @typescript-eslint/ban-types
+      | (string & {})]?: string;
+  }>;
+  canDrag?: DraggableGet<boolean>;
+}
+
+export const useDraggable = <
+  T extends Record<string, unknown> = Record<string, unknown>,
+>(
+  getOptions: () => DraggableOptions<T> = () => ({}),
+  deps: any[] = []
+) => {
+  const [dragging, setDragging] = useState<boolean>(false);
+
+  // used to capture the dragging state when the drag is started
+  const [capturing, setCapturing] = useState<boolean>(false);
```

Comment:
Curious about this state, is it to momentarily apply a special style at the moment drag begins to generate a preview?

Reply:
Yes, when dragging starts, capturing will briefly become true, then become false, and the browser will take a screenshot of the element when capturing = true as a preview of the drag.